src/data/list/cycle.lean

```diff
@@ -0,0 +1,515 @@
+/-
+Copyright (c) 2021 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import data.list.rotate
+import data.finset.basic
+
+/-!
+# Cycles of a list
+
+Lists have an equivalence relation of whether they are rotational permutations of one another.
+This relation is defined as `is_rotated`.
+
+Based on this, we define the quotient of lists by the rotation relation, called `cycle`.
+
+-/
+
+namespace list
+
+variables {α : Type*} [decidable_eq α]
+
+/--
+Given an element `x : α` of `l : list α` such that `x ∈ l`, get the next
+element of `l`. This works from head to tail, (including a check for last element)
+so it will match on first hit, ignoring later duplicates.
+-/
+def next: Π (l : list α) (x : α) (h : x ∈ l), α
+| []             _ h := by simpa using h
+| [y]            _ _ := y
+| (y :: z :: xs) x h := if hx : x = y then z else
+  if x = (y :: z :: xs).last (cons_ne_nil _ _) then y
```

Comment:
Wouldn't this check make `next [1, 2, 3, 4, 2] 2 _` return `1` instead (the first `if` is false, the second `if` is true and we return `1` without going into recursion)? There's also the O(n^2) behaviour of the `last` element being checked in each recursive case.

I would have implemented `next` in two steps like:
```lean
/-- Return the `z` such that `x :: z :: _` appears in `xs`, or `default` if there is no such `z`. -/
def next_or : Π (xs : list α) (x : α) (default : α), α
| [] x default := default
| [y] x default := default -- Handles the not-found and the wraparound case
| (y :: z :: xs) x default := if x = y then z else next_or (z :: xs) x default

def next (xs : list α) (x : α) (h : x ∈ xs) :=
next_or xs x
  (xs.first (ne_nil_of_mem h)) -- or however you write "get first element of nonempty list"
```

Reply:
Yes, but then I do not know how to make the behavior of `prev` do what is expected of it. Do you have thoughts?

Reply:
I pushed commit 51d8fe4 working out the `next_or` proposal concretely. It depends on a new `list.head_of_mem` function that I couldn't easily find yet, although I asked [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there.20code.20for.20X.3F/topic/Getting.20the.20first.20element.20of.20a.20nonempty.20list) if there's a better way to do it.

I didn't change much about the lemmas, only `next_last_cons` needs a new `nodup` assumption (which is available when it gets used).